File: code_samples/section5/problems/problem5_6/problem5_6.cpp
#include <string>
#include <vector>
#include <sstream>
#include <functional>
#include <iostream>
using namespace std;

// ==========================
// BINARY TREE NODE DEFINITION
// ==========================

// Represents a single node in a binary tree
struct TreeNode {
    int val;            // value stored in this node
    TreeNode *left;     // pointer to left child (nullptr if none)
    TreeNode *right;    // pointer to right child (nullptr if none)

    // Constructor initializes value and sets children to nullptr
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

// ==========================
// SERIALIZATION (PREORDER + NULL MARKERS)
// ==========================

// Serializes a binary tree into a single string using preorder traversal:
//
//   node value, then left subtree, then right subtree
//
// Null pointers are written as "#," so the structure can be reconstructed.
// Example (single node 1):
//   "1,#,#,"
string serialize(TreeNode *root) {
    string out; // output string that we will append tokens into

    // DFS function that writes preorder tokens into "out"
    function<void(TreeNode *)> dfs = [&](TreeNode *node) {
        // Null child => write marker and return
        if (!node) {
            out += "#,";
            return;
        }

        // Write the node value followed by a comma delimiter
        out += to_string(node->val) + ",";

        // Serialize left subtree then right subtree
        dfs(node->left);
        dfs(node->right);
    };

    // Start DFS from root
    dfs(root);

    return out;
}

// ==========================
// DESERIALIZATION
// ==========================

// Reconstructs the tree from the serialized string.
//
// Steps:
// 1) Split the string by commas into tokens
// 2) Rebuild the tree using preorder recursion
//    - token "#" => nullptr
//    - number token => create node, then build left, then build right
TreeNode *deserialize(const string &data) {
    // 1) Tokenize the input string by commas
    vector<string> tokens;
    string token;
    stringstream ss(data);

    while (getline(ss, token, ',')) {
        // Skip empty token (happens at end if string ends with comma)
        if (!token.empty()) tokens.push_back(token);
    }

    // idx tracks which token we are currently consuming
    int idx = 0;

    // 2) DFS rebuild function (mirrors serialization order)
    function<TreeNode *()> dfs = [&]() -> TreeNode * {
        // Safety check: if we run out of tokens, return nullptr
        if (idx >= (int)tokens.size()) return nullptr;

        // Read current token and advance
        const string &t = tokens[idx++];

        // "#" represents null pointer
        if (t == "#") return nullptr;

        // Otherwise parse integer and create a node
        int val = stoi(t);
        TreeNode *node = new TreeNode(val);

        // Rebuild left subtree then right subtree (preorder format)
        node->left = dfs();
        node->right = dfs();

        return node;
    };

    // Start rebuild from the first token
    return dfs();
}

// ----------------------------
// Helpers for testing
// ----------------------------

// Convenience function to allocate a node on the heap
TreeNode* node(int v) {
    return new TreeNode(v);
}

// Frees an entire tree using postorder cleanup:
// delete children before deleting the parent.
void freeTree(TreeNode *root) {
    if (!root) return;
    freeTree(root->left);
    freeTree(root->right);
    delete root;
}

// Sideways tree printer
// - Right subtree appears above the node
// - Left subtree appears below the node
void printTreeImpl(TreeNode *root, int indent) {
    if (!root) return;

    // Spaces to indent per level
    const int STEP = 4;

    // Print right subtree first (appears "up")
    printTreeImpl(root->right, indent + STEP);

    // Print current node with indentation
    cout << string(indent, ' ') << root->val << "\n";

    // Print left subtree last (appears "down")
    printTreeImpl(root->left, indent + STEP);
}

// Public wrapper for printing the tree
void printTree(TreeNode *root) {
    cout << "Tree (sideways, right is up):\n";
    printTreeImpl(root, 0);
    cout << "\n";
}

// Round-trip test: root -> serialize -> deserialize -> serialize
// If serialization is correct and deserialization matches it,
// the two serialized strings should match exactly.
void roundTripTest(const string &label, TreeNode *root) {
    cout << "==== " << label << " ====\n";

    // Print original tree
    printTree(root);

    // Serialize original tree
    string s1 = serialize(root);
    cout << "Serialized:   " << s1 << "\n";

    // Deserialize into a new tree
    TreeNode *copy = deserialize(s1);
    cout << "Deserialized tree:\n";
    printTree(copy);

    // Serialize the new tree again
    string s2 = serialize(copy);
    cout << "Re-serialized:" << s2 << "\n";

    // Compare the strings to confirm structure was preserved
    if (s1 == s2) {
        cout << "Round-trip OK (strings match)\n\n";
    } else {
        cout << "Round-trip MISMATCH!\n\n";
    }

    // Cleanup the deserialized tree
    freeTree(copy);
}

// ----------------------------
// Main test harness
// ----------------------------

int main() {
    // 1) Empty tree
    TreeNode *empty = nullptr;

    // 2) Single node tree
    TreeNode *single = node(42);

    // 3) Larger example tree:
    //
    //         1
    //       /   \
    //      2     3
    //     / \   /
    //    4   5 6
    TreeNode *root = node(1);
    root->left = node(2);
    root->right = node(3);
    root->left->left = node(4);
    root->left->right = node(5);
    root->right->left = node(6);

    // Run tests
    roundTripTest("Empty tree", empty);
    roundTripTest("Single-node tree", single);
    roundTripTest("Larger example tree", root);

    // Cleanup original trees
    freeTree(single);
    freeTree(root);
    // empty is nullptr, nothing to free

    return 0;
}

File: code_samples/section5/problems/problem5_6/problem5_6.cs
#nullable enable

using System.Text;

/* ----------------------------
   Tree Printer (sideways)
   ---------------------------- */

// Number of spaces to indent per tree level in the sideways view
const int INDENT = 4;

// Prints a tree sideways for visualization:
// - Right subtree appears above ("up")
// - Left subtree appears below ("down")
void PrintTree(TreeNode? root)
{
    Console.WriteLine("Tree (sideways, right is up):");
    PrintTreeImpl(root, 0);
    Console.WriteLine();
}

// Recursive helper that prints the tree with indentation based on depth
void PrintTreeImpl(TreeNode? node, int indent)
{
    // Base case: nothing to print
    if (node == null) return;

    // Print right subtree first so it appears "up"
    PrintTreeImpl(node.Right, indent + INDENT);

    // Print current node value with indentation
    Console.WriteLine($"{new string(' ', indent)}{node.Val}");

    // Print left subtree last so it appears "down"
    PrintTreeImpl(node.Left, indent + INDENT);
}

/* ----------------------------
   Round-trip test
   ---------------------------- */

// RoundTrip does a "codec round-trip" test:
// 1) Serialize the tree
// 2) Deserialize back into a new tree
// 3) Serialize the new tree again
// 4) Compare the two serialized strings
//
// If the codec is correct (and deterministic), the strings should match.
void RoundTrip(string label, Codec codec, TreeNode? root)
{
    Console.WriteLine($"==== {label} ====");

    // Print original tree
    PrintTree(root);

    // Serialize original tree to a string
    string s1 = Codec.Serialize(root);
    Console.WriteLine($"Serialized:     {s1}");

    // Deserialize into a new tree instance
    TreeNode? copy = Codec.Deserialize(s1);

    // Print reconstructed tree
    Console.WriteLine("Deserialized:");
    PrintTree(copy);

    // Serialize reconstructed tree again
    string s2 = Codec.Serialize(copy);
    Console.WriteLine($"Re-serialized:  {s2}");

    // Compare strings to validate the round-trip
    if (s1 == s2)
        Console.WriteLine("Round-trip OK (strings match)\n");
    else
        Console.WriteLine("Round-trip MISMATCH!\n");
}

/* ----------------------------
   Build test trees
   ---------------------------- */

// 1) Empty tree
TreeNode? empty = null;

// 2) Single node tree
TreeNode single = new(42);

// 3) Larger example tree:
//
//         1
//       /   \
//      2     3
//     / \   /
//    4   5 6
//
TreeNode root = new(1)
{
    Left = new TreeNode(2)
    {
        Left = new TreeNode(4),
        Right = new TreeNode(5)
    },
    Right = new TreeNode(3)
    {
        Left = new TreeNode(6)
    }
};

/* ----------------------------
   Run tests
   ---------------------------- */

// Create codec instance (provides Serialize / Deserialize)
Codec codec = new();

// Run round-trip tests on all sample trees
RoundTrip("Empty tree", codec, empty);
RoundTrip("Single-node tree", codec, single);
RoundTrip("Larger example tree", codec, root);

// ==========================
// DATA STRUCTURES
// ==========================

// Basic binary tree node (nullable children allowed)
class TreeNode(int v)
{
    public int Val = v;         // value stored in this node
    public TreeNode? Left;      // left child (null if none)
    public TreeNode? Right;     // right child (null if none)
}

// ==========================
// CODEC: SERIALIZE / DESERIALIZE
// ==========================

class Codec
{
    // Serializes a binary tree to a string using preorder traversal
    // with null markers.
    //
    // Format:
    // - Each node is written as "value,"
    // - Null pointers are written as "#,"
    //
    // Example (single node 1):
    // "1,#,#,"
    public static string Serialize(TreeNode? root)
    {
        var sb = new StringBuilder();

        // Local DFS helper that appends preorder tokens into sb
        void Dfs(TreeNode? node)
        {
            // Null subtree => write marker
            if (node == null)
            {
                sb.Append("#,");
                return;
            }

            // Write node value followed by comma
            sb.Append(node.Val);
            sb.Append(',');

            // Serialize left then right
            Dfs(node.Left);
            Dfs(node.Right);
        }

        // Start serialization
        Dfs(root);

        return sb.ToString();
    }

    // Deserializes the string produced by Serialize back into a tree.
    //
    // Steps:
    // 1) Split by commas into tokens
    // 2) Rebuild using preorder recursion:
    //    - token "#" => null
    //    - number => create node, build left, then build right
    public static TreeNode? Deserialize(string data)
    {
        // Split into tokens; RemoveEmptyEntries drops the empty token caused
        // by the trailing comma.
        var tokens = new Queue<string>(
            data.Split(',', StringSplitOptions.RemoveEmptyEntries)
        );

        // Local DFS that consumes tokens in preorder order
        TreeNode? Dfs()
        {
            // Safety: no tokens left
            if (tokens.Count == 0) return null;

            // Get next token
            string t = tokens.Dequeue();

            // Null marker => return null pointer
            if (t == "#") return null;

            // Parse node value and create node
            int val = int.Parse(t);
            var node = new TreeNode(val)
            {
                // Rebuild left subtree then right subtree
                Left = Dfs(),
                Right = Dfs()
            };

            return node;
        }

        // Start rebuild from the first token
        return Dfs();
    }
}
